### PR TITLE
[Merged by Bors] - feat(Algebra/Module): presentation of modules

### PR DESCRIPTION
**Pull request by joelriou**
This PR defines a basic API for the description of modules in terms of generators and relations.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/Presentation/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    Submodule.span A (Set.range relations.relation) ≤ LinearMap.ker solution.π := by
+  rw [Submodule.span_le]
+  rintro _ ⟨r, rfl⟩
+  simp only [SetLike.mem_coe, LinearMap.mem_ker, π_relation]
+
+/-- Given `relations : Relations A` and `solution : relations.Solution M`, this is
+the canonical linear map `relations.Quotient →ₗ[A] M` from the module. -/
+noncomputable def fromQuotient : relations.Quotient →ₗ[A] M :=
+  Submodule.liftQ _ solution.π solution.span_relation_le_ker_π
+
```
Comment (erdOne) [inline Mathlib/Algebra/Module/Presentation/Basic.lean:141]: Should there be a (simp) lemma `solution.fromQuotient (Submodule.Quotient.mk x) = solution.π x`?

File: Mathlib/Algebra/Module/Presentation/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+/-- If `M` admits a presentation by generators and relations, then
+linear maps from `M` can be (naturally) identified to the solutions of
+certain linear equations. -/
+@[simps]
+noncomputable def linearMapEquiv : (M →ₗ[A] N) ≃ relations.Solution N where
+  toFun f := solution.postcomp f
+  invFun s := h.desc s
+  left_inv f := h.postcomp_injective (by aesop)
+  right_inv s := by aesop
+
```
Comment (erdOne) [inline Mathlib/Algebra/Module/Presentation/Basic.lean:250]: Can we get the exact sequence $R^R \to R^G \to M \to 0$ given an `IsPresentation`?

File: Mathlib/Algebra/Module/Presentation/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+variable {A} (M : Type v) [AddCommGroup M] [Module A M] (relations : Relations A)
+
+/-- Given an `A`-module `M`, a term in this type is a presentation by `M` by
+generators and relations, where the shapes of the relations are given by
+a certain `relations : Relations A`. -/
+structure Presentation where
+  /-- a solution to the given equations -/
+  solution : relations.Solution M
+  isPresentation : solution.IsPresentation
```
Comment (erdOne) [inline Mathlib/Algebra/Module/Presentation/Basic.lean:282]: My first thought would be to bundle the `relations : Relations A` as well. What are the scenarios where we want a presentation with prescribed relations?

File: Mathlib/Algebra/Module/Presentation/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+section
+
+variable (π : (relations.G →₀ A) →ₗ[A] M) (hπ : ∀ (r : relations.R), π (relations.relation r) = 0)
+
+/-- Given `relations : Relations A` and an `A`-module `M`, this is a constructor
+for `relations.Solution M` for which the data is given as
+a linear map `π : (relations.G →₀ A) →ₗ[A] M`. (See also `ofπ'` for an alternate
+vanishing criterion.) -/
+noncomputable def ofπ : relations.Solution M where
```
Comment (erdOne) [inline Mathlib/Algebra/Module/Presentation/Basic.lean:168]:
```suggestion
@[simps]
noncomputable def ofπ : relations.Solution M where
```
Do we want this simp lemma?

**Review by erdOne — COMMENTED** [review]
Looks nice!
Some feature request & design choice questions.

File: Mathlib/Algebra/Module/Presentation/Basic.lean (hunk 3 above)
Comment (joelriou) [reply Mathlib/Algebra/Module/Presentation/Basic.lean:282]: Yes, I think you are right! I am updating the code.

File: Mathlib/Algebra/Module/Presentation/Basic.lean (hunk 4 above)
Comment (joelriou) [reply Mathlib/Algebra/Module/Presentation/Basic.lean:168]: I would think we do not want this one, but I shall add `@[simps (config := .lemmasOnly)]`.

File: Mathlib/Algebra/Module/Presentation/Basic.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/Algebra/Module/Presentation/Basic.lean:141]: Yes. Indeed, it makes unnecessary the lemma:
```lean
lemma fromQuotient_single (g : relations.G) :
    solution.fromQuotient (Submodule.Quotient.mk (Finsupp.single g 1)) =
      solution.var g := by
  simp
```

File: Mathlib/Algebra/Module/Presentation/Basic.lean (hunk 2 above)
Comment (joelriou) [reply Mathlib/Algebra/Module/Presentation/Basic.lean:292]: Thanks for the suggestion. I have just added this.

**Comment by joelriou** [comment]
Thanks very much @erdOne for the review!

File: Mathlib/Algebra/Module/Presentation/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+/-- Given an `A`-module `M`, a term in this type is a presentation by `M` by
+generators and relations. -/
+@[nolint checkUnivs]
+structure Presentation where
+  /-- the shape of the relations -/
+  relations : Relations A
+  /-- a solution to the given equations -/
+  solution : relations.Solution M
+  isPresentation : solution.IsPresentation
```
Comment (erdOne) [inline Mathlib/Algebra/Module/Presentation/Basic.lean:326]: If you implement it via
```lean
/-- Given an `A`-module `M`, a term in this type is a presentation by `M` by
generators and relations. -/
@[nolint checkUnivs]
structure Presentation extends Relations A, toRelations.Solution M, toSolution.IsPresentation where
```
Then you can directly write
```
example : Presentation A M where
  G := _
  R := _
  relation := _
  var := _
  relation := _
  bijective := _
```
when defining one.
(There are two fields named `relation` so you might need to rename one)

I have no preferences either way though. It's up to you.

File: Mathlib/Algebra/Module/Presentation/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+variable (in `M`) and each `r : R` specifies a linear relation that these
+variables should satisfy. This way, we get a type `relations.Solution M`.
+Then, if `solution : relations.Solution M`, we introduce the predicate
+`solution.IsPresentation` which asserts that `solution` is the universal
+solution to the given equations, i.e. `solution` gives a presentation
+of `M` by generators and relations.
+
+Given `M` and `relations : Relations A`, we also introduce a structure
+`Presentation M relations` which contains a solution of `relations` in `M`
+which is a presentation of `M` by generators and relations.
```
Comment (erdOne) [inline Mathlib/Algebra/Module/Presentation/Basic.lean:31]: This seems outdated.

Comment (joelriou) [reply Mathlib/Algebra/Module/Presentation/Basic.lean]: Thanks!

File: Mathlib/Algebra/Module/Presentation/Basic.lean (hunk 5 above)
Comment (joelriou) [reply Mathlib/Algebra/Module/Presentation/Basic.lean:328]: Thanks for the suggestion. I have expanded the name of `Solution.relation`.

**Comment by erdOne** [comment]
LGTM
maintainer merge

File: Mathlib/Algebra/Module/Presentation/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+namespace Solution
+
+variable {relations M}
+
+section
+
+variable (solution : relations.Solution M)
+
+/-- Given `relations : Relations A` and a solution in `relations.Solution M`, this is
+the linear map `(relations.G →₀ A) →ₗ[A] M` canonically associated to the . -/
```
Comment (jcommelin) [inline Mathlib/Algebra/Module/Presentation/Basic.lean:118]: canonically associated with what?

**Review by jcommelin — COMMENTED** [review]
Thanks! Left 1 comment

bors d+

File: Mathlib/Algebra/Module/Presentation/Basic.lean (hunk 7 above)
Comment (joelriou) [reply Mathlib/Algebra/Module/Presentation/Basic.lean]: Thanks!

**Pull request closed without merging**